### ISSUE TITLE
`pysgpp`: `ImportError: attempted relative import with no known parent package` in colab

### ISSUE THREAD

**Issue by freifrauvonbleifrei** [SGpp/SGpp#184]
As per @MichaelReh 's request I am trying to run SGpp in https://colab.research.google.com . 

Set-up as always, with `!` prepended for shell commands: 
```
!apt-get install scons swig
!wget https://github.com/SGpp/SGpp/archive/v3.2.0.tar.gz
!tar -xvf v3.2.0.tar.gz
!cd SGpp-3.2.0/  && scons -j16 SG_BASE=1 SG_COMBIGRID=1 SG_DATADRIVEN=1 SG_MISC=0 SG_OPTIMIZATION=1 SG_PDE=1 SG_QUADRATURE=1 SG_SOLVER=1 SG_JAVA=0 SG_MATLAB=0 SG_PYTHON=1 COMPILE_BOOST_TESTS=0 RUN_BOOST_TESTS=0
!cd SGpp-3.2.0/  && chmod +x setup.py && python3 ./setup.py install
```
the contents of the install directory are
```
!ls -l /usr/local/lib/python3.6/dist-packages/pysgpp-1.0.0-py3.6.egg/pysgpp
total 57832
drwxr-sr-x 6 root staff     4096 Nov  7 14:31 extensions
-rw-r--r-- 1 root staff      229 Nov  7 14:31 __init__.py
drwxr-sr-x 2 root staff     4096 Nov  7 14:31 __pycache__
-rw-r--r-- 1 root staff  1077109 Nov  7 14:31 pysgpp_swig.py
-rwxr-xr-x 1 root staff 58128584 Nov  7 14:31 _pysgpp_swig.so
```
but now, `import pysgpp` throws the following:
```
ImportError                               Traceback (most recent call last)

<ipython-input-7-a09bffa9f75b> in <module>()
----> 1 import pysgpp
      2 from itertools import product, combinations, permutations,    combinations_with_replacement
      3 from pysgpp.extensions.datadriven.uq.dists import J, Beta, Uniform
      4 from pysgpp.extensions.datadriven.uq.plot.colors import initialize_plotting_style,     load_color, load_font_properties, savefig
      5 from pysgpp.extensions.datadriven.uq.plot.plot3d import plotSG3d

1 frames

/usr/local/lib/python3.6/dist-packages/pysgpp-1.0.0-py3.6.egg/pysgpp/pysgpp_swig.py in <module>()
      5 # the SWIG interface file instead.
      6 
----> 7 from . import _pysgpp_swig
      8 try:
      9     _swig_property = property

ImportError: attempted relative import with no known parent package
```
Googling suggests it might be a problem with Python 3.6. Using other versions would require to run the stuff locally, which means we are back to the long-install-on-your-own-hardware problem. 

Does anyone have smart ideas how to solve this? Maybe it's a different problem altogether? Did you experience this issue with 3.6 before?

**Comment by valentjn** [comment]
@freifrauvonbleifrei I haven't seen this. It could be a notebook thing or a `setup.py` thing (both of which I've rarely used). SG++ is supposed to be working with all Python 3.x versions. So we should not need to change the version.

First step would be reproducing this on own (non-Google) machines. Probably first on the terminal, without notebooks. If this works and your Python is >= 3.6, then it's a notebook thing.

**Comment by hirschsn** [comment]
This might be easy to fix by a simple post-install patch. Try replacing
```py
from . import _pysgpp_swig
```
with:
```py
try:
    from . import _pysgpp_swig
except:
    import _pysgpp_swig
```

Automation:
```bash
sed -ie 's/^\(from . import _pysgpp_swig\)/try:\n    \1\nexcept:\n    import _pysgpp_swig/' file
```

**Comment by freifrauvonbleifrei** [comment]
Hm, by not `pip`/`setup.py` installing it, but setting the path variable from within python I managed to get rid of this error... but now the shared (swigged) library files cannot be found.

I have tried adding the sgpp library path to both `LD_LIBRARY_PATH` and the sys.path variable, but still: 

```
ImportError                               Traceback (most recent call last)

/content/SGpp/lib/pysgpp/pysgpp_swig.py in <module>()
      7 try:
----> 8     from . import _pysgpp_swig
      9 except:

ImportError: libsgppdatadriven.so: cannot open shared object file: No such file or directory
```
even though
```
!ls -l $LD_LIBRARY_PATH
total 215732
-rwxr-xr-x 1 root root 75333472 Nov 14 11:24 libsgppbase.so
-rwxr-xr-x 1 root root 35863688 Nov 14 11:35 libsgppcombigrid.so
-rwxr-xr-x 1 root root 55315272 Nov 14 11:42 libsgppdatadriven.so
[...]
```
and `sys.path` contains it as well: `['/content/SGpp/lib/sgpp', '', '/env/python', '/usr/lib/python36.zip', '/usr/lib/python3.6', '/usr/lib/python3.6/lib-dynload', '/usr/local/lib/python3.6/dist-packages', '/usr/lib/python3/dist-packages', '/usr/local/lib/python3.6/dist-packages/IPython/extensions', '/root/.ipython', '/content/SGpp/lib', '/content/SGpp/lib/pysgpp']`

Also, I have now switched to pulling a github commit, but it's still the same behaviour.

**Comment by valentjn** [comment]
Oh, that’s a known problem. The LD_LIBRARY_PATH has to be set before starting the notebook. But we don’t have control over that, so I don’t think you can get around this.

**Comment by freifrauvonbleifrei** [comment]
Hm, I have tried that here now
```
from ctypes import *
cdll.LoadLibrary(os.path.join(sgpp_path, "libsgppbase.so")) 
cdll.LoadLibrary(os.path.join(sgpp_path, "libsgppoptimization.so")) 
```
which works for base, but for optimization, base is not found again: `OSError: libsgppbase.so: cannot open shared object file: No such file or directory`

I would really like to cheat my way around this! ;)

**Comment by hirschsn** [comment]
@freifrauvonbleifrei just talked to @leiterrl about it. He has the same problem. Maybe speak to him. :)
tl;dr: I suggest just installing sgpp under /usr as you seem to have root privileges in this VM.

**Comment by freifrauvonbleifrei** [comment]
yup, @leiterrl helped me with that. Now this here works! 
```
!apt-get install scons swig
! git clone https://github.com/SGpp/SGpp.git
! cd SGpp && git checkout b0c10d35c50f147f17df958ff37152b3a13e901a
# install sgpp
!cd SGpp  && scons -j16 SG_BASE=1 SG_COMBIGRID=1 SG_DATADRIVEN=1 SG_MISC=0 SG_OPTIMIZATION=1 SG_PDE=1 SG_QUADRATURE=1 SG_SOLVER=1 SG_JAVA=0 SG_MATLAB=0 SG_PYTHON=1 COMPILE_BOOST_TESTS=0 RUN_BOOST_TESTS=0
import sys
import os

sgpp_folder = "SGpp"

# Make python interface available
if not 'moduleDir' in globals():
    moduleDir = os.path.join(os.path.join(os.getcwd(), sgpp_folder),"lib")
    moduleDir2 = os.path.join(os.path.join(os.getcwd(), sgpp_folder),"lib/pysgpp")

sys.path.append(moduleDir)
sys.path.append(moduleDir2)
# print(sys.path)

sgpp_path = os.path.join(os.path.join(os.getcwd(), sgpp_folder),"lib/sgpp")
!sed -ie 's/^\(from . import extensions\)/try:\n    \1\nexcept:\n    import extensions/' /content/SGpp/lib/pysgpp/__init__.py
!cp $sgpp_path/* /usr/lib
import pysgpp
```
I will tidy it up and make available to @MichaelReh ;)

**Comment by valentjn** [comment]
Great! But I suggest making this available to everyone. In the long term, a “try SG++ here” would be great, but in the meantime, this should be in the Wiki, shouldn’t it?

**Comment by MichaelReh** [comment]
Great. Thank you very much for setting this up on such short notice! I really appreciate it.

**Comment by freifrauvonbleifrei** [comment]
here goes:
https://github.com/SGpp/SGpp/wiki/Installing-pysgpp-on-colab

actually I wanted to upload the `.ipynb` file, but seems I today am to stupid to upload files in the wiki :/

**Comment by hirschsn** [comment]
@freifrauvonbleifrei 

> `!cp $sgpp_path/* /usr/lib`

Always quote $ expansions with unknown contents.

**Comment by MichaelReh** [comment]
@freifrauvonbleifrei 
Why did you hard code the used version of SG++ to commit b0c10d35c50f147f17df958ff37152b3a13e901a?

Was there a problem with simply checking out the latest master branch?
(I just tried that and it worked in my Colab instance with no problems.)

**Comment by freifrauvonbleifrei** [comment]
@valentjn no, there should be no problem -- I just wanted to have a definite reference to a branch that works currently, in case something breaks in the future (then we know whether it's due to colab or SGpp changes).

**Comment by valentjn** [comment]
@freifrauvonbleifrei Huh? I think you meant to tag @MichaelReh ;)

**Comment by freifrauvonbleifrei** [comment]
@valentjn @MichaelReh yup!
